File: src/main/java/com/ihd2/Sodasumo.java
package com.ihd2;

import com.ihd2.model.Model;
import org.xml.sax.SAXException;

import javax.swing.*;
import java.awt.*;
import java.awt.event.*;
import java.io.File;
import java.io.FilenameFilter;

public class Sodasumo extends JFrame implements MouseListener, ItemListener, ActionListener, KeyListener {

    private static final String VERSION = "SodaSumo 1.6";

    private static final Dimension GAME_DIMENSION = new Dimension(1000, 350);
    private static final Dimension CP_DIMENSION = new Dimension(150, 320);
    public static double GAME_WIDTH = GAME_DIMENSION.getWidth() - CP_DIMENSION.getWidth();
    private final JButton loadButton;
    private final JButton stopButton;
    private final JCheckBox invertM1Button;
    private final JCheckBox invertM2Button;
    private final GameDraw newGameDraw = new GameDraw();
    private final JMenuItem aboutItemSoda;
    private final JMenuItem aboutItemAuthor;
    private String[] xmlFiles;
    private boolean invertM1 = false, invertM2 = false;
    private final JComboBox<Object> box1;
    private final JComboBox<Object> box2;
    private final JComboBox<String> boxTime;
    private XmlParser xp1, xp2;

    private Sodasumo() {
        loadDirectory();
        JMenuBar menuBar = new JMenuBar();
        this.setJMenuBar(menuBar);
        menuBar.setBackground(Color.GRAY.brighter().brighter());
        setTitle(VERSION);

        JMenu aboutMenu = new JMenu("About");
        menuBar.add(aboutMenu);

        aboutItemSoda = new JMenuItem(VERSION);
        aboutMenu.add(aboutItemSoda);
        aboutItemSoda.addMouseListener(this);
        aboutItemAuthor = new JMenuItem("Author");
        aboutMenu.add(aboutItemAuthor);
        aboutItemAuthor.addMouseListener(this);

        Container contents = getContentPane();
        contents.setLayout(new BorderLayout());
        contents.setBackground(Color.WHITE);
        setSize(GAME_DIMENSION);
        setResizable(false);
        contents.add(newGameDraw);
        JPanel cPanel = new JPanel();
        contents.add(cPanel, BorderLayout.EAST);
        cPanel.setBackground(Color.GRAY.brighter());
        cPanel.setPreferredSize(CP_DIMENSION);

        box1 = new JComboBox<>(xmlFiles);
        box1.setMaximumRowCount(10);
        box1.setPrototypeDisplayValue("THIS IS A VERY LON");
        box1.setSelectedItem("daintywalker");
        box1.setEditable(true);
        box1.addActionListener(this);
        cPanel.add(box1);

        invertM1Button = new JCheckBox("Invert Direction", false);
        invertM1Button.setBackground(Color.GRAY.brighter());
        invertM1Button.addItemListener(this);
        cPanel.add(invertM1Button);

        box2 = new JComboBox<>(xmlFiles);
        box2.setMaximumRowCount(10);
        box2.setPrototypeDisplayValue("THIS IS A VERY LON");
        box2.setSelectedItem("KeniF_triangle");
        box2.setEditable(true);
        box2.addActionListener(this);
        cPanel.add(box2);

        invertM2Button = new JCheckBox("Invert Direction", false);
        invertM2Button.setBackground(Color.GRAY.brighter());
        invertM2Button.addItemListener(this);
        cPanel.add(invertM2Button);

        loadButton = new JButton("  Start  ");
        loadButton.addMouseListener(this);
        loadButton.addKeyListener(this);
        loadButton.addActionListener(this);
        cPanel.add(loadButton);

        stopButton = new JButton("  Stop  ");
        stopButton.addMouseListener(this);
        stopButton.addKeyListener(this);
        stopButton.setVisible(false);
        stopButton.addActionListener(this);
        cPanel.add(stopButton);

        JLabel emptySpace = new JLabel("                                    ");
        JLabel emptySpace2 = new JLabel("                                    ");
        JLabel eS3 = new JLabel("                                    ");
        JLabel label = new JLabel("    Time Limit");
        cPanel.add(emptySpace);
        cPanel.add(emptySpace2);
        cPanel.add(eS3);
        cPanel.add(label);

        String[] times = {"5", "10", "15", "20", "25", "30", "60"};
        boxTime = new JComboBox<>(times);
        boxTime.setMaximumRowCount(10);

        boxTime.setSelectedItem("15");
        boxTime.setEditable(false);
        boxTime.addActionListener(this);
        boxTime.setToolTipText("Set Time Limit");
        boxTime.setPrototypeDisplayValue("30");
        cPanel.add(boxTime);


        setDefaultCloseOperation(EXIT_ON_CLOSE);
        setLocationRelativeTo(null);

    }

    public static void main(String[] args) {
        Sodasumo newGame = new Sodasumo();
        newGame.setVisible(true);
    }

    private void loadDirectory() {
        try {
            String currentDir = System.getProperty("user.dir");
            File dir = new File(currentDir);
            FilenameFilter filter = (dir1, name) -> name.endsWith(".xml");
            xmlFiles = dir.list(filter);
            if (xmlFiles.length == 0) throw new Exception();
            for (int i = 0; i < xmlFiles.length; i++) {
                xmlFiles[i] = xmlFiles[i].substring(0, xmlFiles[i].length() - 4);
            }
        } catch (Exception e) {
            JOptionPane.showMessageDialog(this, "No XML files in directory!!\n",
                    "Error", JOptionPane.WARNING_MESSAGE);
        }
    }

    public void itemStateChanged(ItemEvent e) {
        ItemSelectable source = e.getItemSelectable();
        if (source == invertM1Button) {
            invertM1 = !invertM1;
            newGameDraw.invertM1();
        } else if (source == invertM2Button) {
            invertM2 = !invertM2;
            newGameDraw.invertM2();
        }
    }

    public void actionPerformed(ActionEvent e) {
        if (e.getSource() == box1) {
            Object b = box1.getSelectedItem();
            if (box1.getSelectedIndex() == -1 && b.toString().contains("/")) {
                box1.addItem(b);
                box2.addItem(b);
            }
            box1.showPopup();
        } else if (e.getSource() == box2) {
            Object a = box2.getSelectedItem();
            if (box2.getSelectedIndex() == -1 && a.toString().contains("/")) {
                box2.addItem(a);
                box1.addItem(a);
            }
            box2.showPopup();
        } else if (e.getSource() == boxTime) {
            newGameDraw.setTimeLimit(Double.parseDouble(boxTime.getSelectedItem().toString()) * 1000.0);
        }
    }

    public void mouseExited(MouseEvent e) {
    }

    public void mouseClicked(MouseEvent e) {
    }

    public void mouseEntered(MouseEvent e) {
    }

    public void keyTyped(KeyEvent e) {
    }

    public void keyPressed(KeyEvent e) {
        try {
            if ((e.getKeyCode() == KeyEvent.VK_ENTER ||
                    e.getKeyCode() == KeyEvent.VK_SPACE) && loadButton.isVisible()) {
                String xmlFile1, xmlFile2;
                xmlFile1 = box1.getSelectedItem().toString();
                xp1 = new XmlParser(xmlFile1 + ".xml");
                if (xp1.getVerified() != 2)
                    throw new Exception("Model 1 cannot be verified!");
                xmlFile2 = box2.getSelectedItem().toString();
                xp2 = new XmlParser(xmlFile2 + ".xml");
                if (xp2.getVerified() != 2)
                    throw new Exception("Model 2 cannot be verified!");
                Model model1 = xp1.getModel();
                model1.setName(box1.getSelectedItem().toString());
                Model model2 = xp2.getModel();
                model2.setName(box2.getSelectedItem().toString());
                newGameDraw.provideModel1(model1);
                newGameDraw.provideModel2(model2);
                loadButton.setVisible(false);
                stopButton.setVisible(true);
                stopButton.grabFocus();
                box1.setEnabled(false);
                box2.setEnabled(false);
                newGameDraw.startDraw();
                newGameDraw.init();
            } else if ((e.getKeyCode() == KeyEvent.VK_ENTER ||
                    e.getKeyCode() == KeyEvent.VK_SPACE) && stopButton.isVisible()) {
                newGameDraw.pause();
                stopButton.setVisible(false);
                loadButton.setVisible(true);
                box1.setEnabled(true);
                loadButton.grabFocus();
                box2.setEnabled(true);
            }
        } catch (java.io.IOException f) {
            JOptionPane.showMessageDialog(newGameDraw, "Please ensure the XML files are in the same directory.\n" + f,
                    "Error During Load", JOptionPane.WARNING_MESSAGE);
        } catch (SAXException s) {
            JOptionPane.showMessageDialog(newGameDraw, "Invalid XML files.\n" + s,
                    "Error During Load", JOptionPane.WARNING_MESSAGE);
        } catch (java.lang.NullPointerException ed) {
            JOptionPane.showMessageDialog(newGameDraw, "Model contains objects not permitted in SodaSumo\n" + ed,
                    "Error During Load", JOptionPane.WARNING_MESSAGE);
        } catch (Exception ds) {
            JOptionPane.showMessageDialog(newGameDraw, "Error when loading models\n" + ds,
                    "Error During Load", JOptionPane.WARNING_MESSAGE);
        }

    }

    public void keyReleased(KeyEvent e) {
    }

    public void mousePressed(MouseEvent e) {
        try {
            String xmlFile1, xmlFile2;
            if (e.getSource() == loadButton) {
                xmlFile1 = box1.getSelectedItem().toString();
                xp1 = new XmlParser(xmlFile1 + ".xml");
                if (xp1.getVerified() != 2)
                    throw new Exception("Model 1 cannot be verified!");
                xmlFile2 = box2.getSelectedItem().toString();
                xp2 = new XmlParser(xmlFile2 + ".xml");
                if (xp2.getVerified() != 2)
                    throw new Exception("Model 2 cannot be verified!");
                Model model1 = xp1.getModel();
                model1.setName(box1.getSelectedItem().toString());
                Model model2 = xp2.getModel();
                model2.setName(box2.getSelectedItem().toString());
                newGameDraw.provideModel1(model1);
                newGameDraw.provideModel2(model2);
                loadButton.setVisible(false);
                stopButton.setVisible(true);
                box1.setEnabled(false);
                box2.setEnabled(false);
                newGameDraw.startDraw();
                newGameDraw.init();
            } else if (e.getSource() == stopButton) {
                newGameDraw.pause();
                loadButton.setVisible(true);
                stopButton.setVisible(false);
                box1.setEnabled(true);
                box2.setEnabled(true);
            } else if (e.getSource() == aboutItemSoda) {
                JOptionPane.showMessageDialog(newGameDraw,
                        "<html><font size=5 color=blue><b><u>" + VERSION + "</u></b></font></html>\n\n" +
                                "SodaSumo is a clone/extension to the physics game SodaRace.\n" +
                                "While SodaRace allows users to import several models designed\n" +
                                "in SodaConstructor for a competition in moving speed,\n" +
                                "SodaSumo replicates the SodaConstructor physics engine\n" +
                                "and takes the game further by implementing collision detection\n" +
                                "between two models.\n" +
                                "\nMore information at\n" +
                                "<html><a href=\"http://sodaplay.com\">www.sodaplay.com</a></html>",
                        "About", JOptionPane.INFORMATION_MESSAGE);
            } else if (e.getSource() == aboutItemAuthor) {
                JOptionPane.showMessageDialog(newGameDraw,
                        "<html><font size=5 color=blue><b><u>Who made SodaSumo?</u></b></font></html>\n\n" +
                                "(C) 2008, 2020 Kenneth \"KeniF\" Lam | kenif.lam@gmail.com\n\n" +
                                "This program was started as my third year project\n" +
                                "at the University of Manchester, United Kingdom.\n" +
                                "Supervised by Dr. Mary McGee Wood",
                        "About", JOptionPane.INFORMATION_MESSAGE);
            }

        } catch (java.io.IOException f) {
            JOptionPane.showMessageDialog(newGameDraw, "Please ensure the XML files are in the same directory\n" + f,
                    "Error During Load", JOptionPane.WARNING_MESSAGE);
        } catch (SAXException s) {
            JOptionPane.showMessageDialog(newGameDraw, "Invalid XML files.\n" + s,
                    "Error During Load", JOptionPane.WARNING_MESSAGE);
        } catch (java.lang.NullPointerException es) {
            JOptionPane.showMessageDialog(newGameDraw, "Model contains objects not permitted in SodaSumo\n" + es,
                    "Error During Load", JOptionPane.WARNING_MESSAGE);
        } catch (Exception ds) {
            JOptionPane.showMessageDialog(newGameDraw, "Error when loading models\n" + ds,
                    "Error During Load", JOptionPane.WARNING_MESSAGE);
        }
    }

    public void mouseReleased(MouseEvent e) {
    }
}

File: src/main/kotlin/com/ihd2/Sodasumo.kt
package com.ihd2

import java.io.FilenameFilter
import java.io.IOException
import java.lang.NullPointerException
import org.xml.sax.SAXException
import java.awt.Dimension
import kotlin.jvm.JvmStatic
import java.awt.Color
import java.awt.BorderLayout
import java.awt.event.*
import java.io.File
import java.lang.Exception
import javax.swing.*

class Sodasumo private constructor() : JFrame(), MouseListener, ItemListener, ActionListener, KeyListener {
    private val loadButton: JButton
    private val stopButton: JButton
    private val stepButton: JButton
    private val pauseButton: JButton
    private val invertM1Button: JCheckBox
    private val invertM2Button: JCheckBox
    private val gameDraw = GameDraw()
    private val aboutItemSoda: JMenuItem
    private val aboutItemAuthor: JMenuItem
    private var xmlFiles: Array<String> = arrayOf()
    private var invertM1 = false
    private var invertM2 = false
    private val box1: JComboBox<Any>
    private val box2: JComboBox<Any>
    private val boxTime: JComboBox<String>
    private var xp1: XmlParser? = null
    private var xp2: XmlParser? = null
    private fun loadDirectory() {
        try {
            val currentDir = System.getProperty("user.dir")
            val dir = File(currentDir)
            val filter = FilenameFilter { _, name: String -> name.endsWith(".xml") }
            xmlFiles = dir.list(filter)
            if (xmlFiles.isEmpty()) throw Exception()
            for (i in xmlFiles.indices) {
                xmlFiles[i] = xmlFiles[i].substring(0, xmlFiles[i].length - 4)
            }
        } catch (e: Exception) {
            JOptionPane.showMessageDialog(
                this, "No XML files in directory!!\n",
                "Error", JOptionPane.WARNING_MESSAGE
            )
        }
    }

    override fun itemStateChanged(e: ItemEvent) {
        val source = e.itemSelectable
        if (source === invertM1Button) {
            invertM1 = !invertM1
            gameDraw.invertM1()
        } else if (source === invertM2Button) {
            invertM2 = !invertM2
            gameDraw.invertM2()
        }
    }

    override fun actionPerformed(e: ActionEvent) {
        when {
            e.source === box1 -> box1.showPopup()
            e.source === box2 -> box2.showPopup()
            e.source === boxTime -> gameDraw.setTimeLimit(
                (boxTime.selectedItem?.toString()?.toLong() ?: 0) * 1000L)
        }
    }

    override fun mouseExited(e: MouseEvent) {}
    override fun mouseClicked(e: MouseEvent) {}
    override fun mouseEntered(e: MouseEvent) {}
    override fun keyTyped(e: KeyEvent) {}
    override fun keyPressed(e: KeyEvent) {
        try {
            if ((e.keyCode == KeyEvent.VK_ENTER || e.keyCode == KeyEvent.VK_SPACE)) {
                when {
                    e.source === loadButton -> loadAndStart()
                    e.source === stopButton -> stop()
                    e.source === pauseButton -> pause()
                    e.source === stepButton -> stepAnimation()
                }
            }
        } catch (f: IOException) {
            JOptionPane.showMessageDialog(
                gameDraw, "Please ensure the XML files are in the same directory.\n$f",
                "Error During Load", JOptionPane.WARNING_MESSAGE
            )
        } catch (s: SAXException) {
            JOptionPane.showMessageDialog(
                gameDraw, "Invalid XML files.\n$s",
                "Error During Load", JOptionPane.WARNING_MESSAGE
            )
        } catch (ed: NullPointerException) {
            JOptionPane.showMessageDialog(
                gameDraw, "Model contains objects not permitted in SodaSumo\n$ed",
                "Error During Load", JOptionPane.WARNING_MESSAGE
            )
        } catch (ds: Exception) {
            JOptionPane.showMessageDialog(
                gameDraw, "Error when loading models\n$ds",
                "Error During Load", JOptionPane.WARNING_MESSAGE
            )
        }
    }

    private fun stop() {
        gameDraw.stop()
        stopButton.isVisible = false
        loadButton.isVisible = true
        box1.isEnabled = true
        loadButton.grabFocus()
        box2.isEnabled = true
    }

    private fun loadAndStart() {
        val xmlFile1 = box1.selectedItem?.toString() ?: ""
        xp1 = XmlParser("$xmlFile1.xml")
        if (xp1!!.verified != 2) throw Exception("Model 1 cannot be verified!")
        val xmlFile2 = box2.selectedItem?.toString() ?: ""
        xp2 = XmlParser("$xmlFile2.xml")
        if (xp2!!.verified != 2) throw Exception("Model 2 cannot be verified!")
        val model1 = xp1!!.model
        model1.name = box1.selectedItem?.toString() ?: ""
        val model2 = xp2!!.model
        model2.name = box2.selectedItem?.toString() ?: ""
        gameDraw.provideModel1(model1)
        gameDraw.provideModel2(model2)
        loadButton.isVisible = false
        stopButton.isVisible = true
        box1.isEnabled = false
        box2.isEnabled = false
        gameDraw.startDraw()
        gameDraw.init()
    }

    override fun keyReleased(e: KeyEvent) {}
    override fun mousePressed(e: MouseEvent) {
        try {
            when {
                e.source === loadButton -> loadAndStart()
                e.source === stopButton -> stop()
                e.source === pauseButton -> pause()
                e.source === stepButton -> stepAnimation()
                e.source === aboutItemSoda -> {
                    JOptionPane.showMessageDialog(
                        gameDraw,
                        """
                                <html><font size=5 color=blue><b><u>$APP_NAME $VERSION</u></b></font></html>
                                
                                SodaSumo is a clone/extension to the physics game SodaRace.
                                While SodaRace allows users to import several models designed
                                in SodaConstructor for a competition in moving speed,
                                SodaSumo replicates the SodaConstructor physics engine
                                and takes the game further by implementing collision detection
                                between two models.
                                
                                More information at
                                <html><a href="http://sodaplay.com">www.sodaplay.com</a></html>
                                """.trimIndent(),
                        "About", JOptionPane.INFORMATION_MESSAGE
                    )
                }
                e.source === aboutItemAuthor -> {
                    JOptionPane.showMessageDialog(
                        gameDraw,
                        """<html><font size=5 color=blue><b><u>Who made SodaSumo?</u></b></font></html>
        
(C) 2008, 2020 Kenneth "KeniF" Lam | kenif.lam@gmail.com
        
This program was started as my third year project
at the University of Manchester, United Kingdom.
Supervised by Dr. Mary McGee Wood""",
                        "About", JOptionPane.INFORMATION_MESSAGE
                    )
                }
            }
        } catch (f: IOException) {
            JOptionPane.showMessageDialog(
                gameDraw, "Please ensure the XML files are in the same directory\n$f",
                "Error During Load", JOptionPane.WARNING_MESSAGE
            )
        } catch (s: SAXException) {
            JOptionPane.showMessageDialog(
                gameDraw, "Invalid XML files.\n$s",
                "Error During Load", JOptionPane.WARNING_MESSAGE
            )
        } catch (es: NullPointerException) {
            JOptionPane.showMessageDialog(
                gameDraw, "Model contains objects not permitted in SodaSumo\n$es",
                "Error During Load", JOptionPane.WARNING_MESSAGE
            )
        } catch (ds: Exception) {
            JOptionPane.showMessageDialog(
                gameDraw, "Error when loading models\n$ds",
                "Error During Load", JOptionPane.WARNING_MESSAGE
            )
        }
    }

    private fun stepAnimation() {
        gameDraw.stepAnimationManually()
    }

    private fun pause() {
        gameDraw.paused = !gameDraw.paused
    }

    override fun mouseReleased(e: MouseEvent) {}

    companion object {
        private val GAME_DIMENSION = Dimension(1000, 350)
        private val SIDE_PANEL_DIMENSION = Dimension(150, 320)

        @JvmStatic
        fun main(args: Array<String>) {
            val newGame = Sodasumo()
            newGame.isVisible = true
        }
    }

    init {
        loadDirectory()
        val menuBar = JMenuBar()
        this.jMenuBar = menuBar
        menuBar.background = Color.GRAY.brighter().brighter()
        title = "$APP_NAME $MAJOR_VERSION"
        val aboutMenu = JMenu("About")
        menuBar.add(aboutMenu)
        aboutItemSoda = JMenuItem("$APP_NAME $VERSION")
        aboutMenu.add(aboutItemSoda)
        aboutItemSoda.addMouseListener(this)
        aboutItemAuthor = JMenuItem("Author")
        aboutMenu.add(aboutItemAuthor)
        aboutItemAuthor.addMouseListener(this)
        val contents = contentPane
        contents.layout = BorderLayout()
        contents.background = Color.WHITE
        size = GAME_DIMENSION
        isResizable = false
        contents.add(gameDraw)
        val cPanel = JPanel()
        contents.add(cPanel, BorderLayout.EAST)
        cPanel.background = Color.GRAY.brighter()
        cPanel.preferredSize = SIDE_PANEL_DIMENSION
        box1 = JComboBox(xmlFiles)
        box1.maximumRowCount = 10
        box1.prototypeDisplayValue = "THIS IS A VERY LON"
        box1.selectedItem = "daintywalker"
        box1.isEditable = true
        box1.addActionListener(this)
        cPanel.add(box1)
        invertM1Button = JCheckBox("Invert Direction", false)
        invertM1Button.background = Color.GRAY.brighter()
        invertM1Button.addItemListener(this)
        cPanel.add(invertM1Button)
        box2 = JComboBox(xmlFiles)
        box2.maximumRowCount = 10
        box2.prototypeDisplayValue = "THIS IS A VERY LON"
        box2.selectedItem = "KeniF_triangle"
        box2.isEditable = true
        box2.addActionListener(this)
        cPanel.add(box2)
        invertM2Button = JCheckBox("Invert Direction", false)
        invertM2Button.background = Color.GRAY.brighter()
        invertM2Button.addItemListener(this)
        cPanel.add(invertM2Button)
        loadButton = JButton("  Start  ")
        loadButton.addMouseListener(this)
        loadButton.addKeyListener(this)
        cPanel.add(loadButton)
        stopButton = JButton("  Stop  ")
        stopButton.addMouseListener(this)
        stopButton.addKeyListener(this)
        stopButton.isVisible = false
        cPanel.add(stopButton)
        cPanel.add(JLabel("                                    "))

        pauseButton = JButton("Pause / Resume")
        pauseButton.addMouseListener(this)
        pauseButton.addKeyListener(this)
        cPanel.add(pauseButton)

        stepButton = JButton("Step")
        stepButton.addMouseListener(this)
        stepButton.addKeyListener(this)
        cPanel.add(stepButton)

        val label = JLabel("   Time Limit")
        cPanel.add(label)

        val times = arrayOf("5", "10", "15", "20", "25", "30", "60")
        boxTime = JComboBox(times)
        boxTime.maximumRowCount = 10
        boxTime.selectedItem = "15"
        boxTime.isEditable = false
        boxTime.addActionListener(this)
        boxTime.toolTipText = "Set Time Limit"
        boxTime.prototypeDisplayValue = "30"
        cPanel.add(boxTime)
        defaultCloseOperation = EXIT_ON_CLOSE
        setLocationRelativeTo(null)
    }
}